Documentation/guides/cpp_cmake.rst

```diff
@@ -0,0 +1,300 @@
+    )
+
+    link_directories(${EXE_NAME} ${NUTTX_PATH}/libs)
+
+    add_executable(${EXE_NAME} ${SOURCE_FILES} main.cpp ${HEADER_FILES})
+
+    add_custom_command(
+            TARGET ${EXE_NAME}
+            POST_BUILD
+            COMMAND ${CMAKE_OBJCOPY} ARGS -S -O binary ${CMAKE_BINARY_DIR}/${EXE_NAME}.elf ${CMAKE_BINARY_DIR}/${EXE_NAME}.bin
+    )
+
+    target_link_libraries(${EXE_NAME} --start-group)
+
+    target_link_libraries(${EXE_NAME} sched)
+    target_link_libraries(${EXE_NAME} drivers)
+    target_link_libraries(${EXE_NAME} boards)
+    target_link_libraries(${EXE_NAME} c)
+    target_link_libraries(${EXE_NAME} mm)
+    target_link_libraries(${EXE_NAME} arch)
+    target_link_libraries(${EXE_NAME} xx)
+    target_link_libraries(${EXE_NAME} apps)
+    target_link_libraries(${EXE_NAME} fs)
+    target_link_libraries(${EXE_NAME} binfmt)
+    target_link_libraries(${EXE_NAME} board)
+    target_link_libraries(${EXE_NAME} gcc)
+    target_link_libraries(${EXE_NAME} supc++)
+
+    target_link_libraries(${EXE_NAME} --end-group)
+
+* hellocpp/src/main.cpp
+
+.. code-block:: c++
+    :language: c++
+    :caption: main.cpp
+    :name: main.cpp
+
+    #include <stdio.h>
+    #include "HelloWorld.h"
+    #include <nuttx/config.h>
+
+    extern "C"
+    {
+            int hellocpp_main(void)
+            {
+
+                    CHelloWorld *pHelloWorld = new CHelloWorld();
+                    pHelloWorld->HelloWorld();
+
+                    CHelloWorld helloWorld;
+                    helloWorld.HelloWorld();
+
+                    delete pHelloWorld;
+                    return 0;
+            }
+    }
+
+* hellocpp/src/HelloWorld.h
+
+.. code-block:: c++
+    :language: c++
+    :caption: HelloWorld.h
+    :name: HelloWorld.h
+
+    #ifndef HELLOWORLD_H_
+    #define HELLOWORLD_H_
+
+    #include "nuttx/config.h"
+
+    class CHelloWorld
+    {
+            public:
+                    CHelloWorld();
+                    ~CHelloWorld();
+                    bool HelloWorld(void);
+            private:
+                    int mSecret;
+    };
+
+    #endif
+
+* hellocpp/src/HelloWorld.cpp
+
+.. code-block:: c++
+    :language: c++
+    :caption: HelloWorld.cpp
+    :name: HelloWorld.cpp
+
+    #include <cstdio>
+    #include <string>
+
+    #include "HelloWorld.h"
+
+    CHelloWorld::CHelloWorld() {
+            mSecret = 42;
+            printf("Constructor: mSecret=%d\n",mSecret);
+    }
+
+    CHelloWorld::~CHelloWorld() {
+
+    }
+
+    bool CHelloWorld::HelloWorld(void) {
+            printf("HelloWorld: mSecret=%d\n",mSecret);
+
+            std::string sentence = "Hello";
+            printf("TEST=%s\n",sentence.c_str());
+
+            if (mSecret == 42) {
+                    printf("CHelloWorld: HelloWorld: Hello, world!");
+                    return true;
+            }
+            else {
+                    printf("CHelloWorld: HelloWorld: CONSTRUCTION FAILED!\n");
+                    return false;
```

Comment:
```suggestion
            std::printf("Constructor: mSecret=%d\n",mSecret);
    }

    CHelloWorld::~CHelloWorld() {

    }

    bool CHelloWorld::HelloWorld(void) {
            std::printf("HelloWorld: mSecret=%d\n",mSecret);

            std::string sentence = "Hello";
            std::printf("TEST=%s\n",sentence.c_str());

            if (mSecret == 42) {
                    std::printf("CHelloWorld: HelloWorld: Hello, world!");
                    return true;
            }
            else {
                    std::printf("CHelloWorld: HelloWorld: CONSTRUCTION FAILED!\n");
                    return false;
```
Sorry being to picky on this, but it may fail to compile due to the missing namespace.

Reply:
Won't it be a problem with NuttX's printf?

Some functions, like printf, seems to be better used not from the language/compiler standard library, but from the platform. That's why I left the printf and stdio.h (and the testcxx example too ...).

In standard C++ project, in general, printf is not used. But I believe in NuttX it is better to use.

Reply:
If the C++ library is correctly configured, it shouldn't be a problem.
`std::prinf` is the C++ STL version, which is basically a wrapper for the C language version.